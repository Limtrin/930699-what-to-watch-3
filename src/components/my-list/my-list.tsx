import React from "react";
import PropTypes from "prop-types";
import {Link} from "react-router-dom";
import {connect} from "react-redux";
import FilmCard from '../film-card/film-card.jsx';
import {getMyListFilms} from "../../reducer/data/selectors.js";
import {Operation as DataOperation} from "../../reducer/data/data.js";
import withVideo from '../../hocs/with-video/with-video.js';

const FilmCardWrapper = withVideo(FilmCard);

class MyList extends React.PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.props.loading();
  }

  render() {
    const {filmsList, onHeaderClickHandler, onFilmCardClickHandler, onItemEnterHandler, onItemLeaveHandler} = this.props;
    return filmsList && (
      <>
        <div className="user-page">
          <header className="page-header user-page__head">
            <div className="logo">
              <Link to="/" className="logo__link">
                <span className="logo__letter logo__letter--1">W</span>
                <span className="logo__letter logo__letter--2">T</span>
                <span className="logo__letter logo__letter--3">W</span>
              </Link>
            </div>

            <h1 className="page-title user-page__title">My list</h1>

            <div className="user-block">
              <div className="user-block__avatar">
                <img src="img/avatar.jpg" alt="User avatar" width="63" height="63" />
              </div>
            </div>
          </header>

          <section className="catalog">
            <h2 className="catalog__title visually-hidden">Catalog</h2>

            <div className="catalog__movies-list">
              {filmsList.map((item) => (
                <FilmCardWrapper
                  key={item.id}
                  film={item}
                  onHeaderClickHandler={onHeaderClickHandler}
                  onFilmCardClickHandler={onFilmCardClickHandler}
                  onCardMouseEnterHandler={onItemEnterHandler}
                  onCardMouseLeaveHandler={onItemLeaveHandler}
                />
              ))}
            </div>

          </section>

          <footer className="page-footer">
            <div className="logo">
              <a className="logo__link logo__link--light">
                <span className="logo__letter logo__letter--1">W</span>
                <span className="logo__letter logo__letter--2">T</span>
                <span className="logo__letter logo__letter--3">W</span>
              </a>
            </div>

            <div className="copyright">
              <p>Â© 2019 What to watch Ltd.</p>
            </div>
          </footer>
        </div>
      </>
    );
  }
}

MyList.propTypes = {
  onFilmFavoriteStatusClickHandler: PropTypes.func.isRequired,
  filmsList: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    year: PropTypes.number.isRequired,
    image: PropTypes.string.isRequired,
    poster: PropTypes.string.isRequired,
    cover: PropTypes.string.isRequired,
    preview: PropTypes.string.isRequired,
    time: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    votes: PropTypes.number.isRequired,
    director: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    reviews: PropTypes.arrayOf(
        PropTypes.shape({
          rating: PropTypes.number.isRequired,
          date: PropTypes.string.isRequired,
          author: PropTypes.shape({
            id: PropTypes.number.isRequired,
            name: PropTypes.string.isRequired,
          }).isRequired,
          text: PropTypes.string.isRequired
        }).isRequired
    ),
    starring: PropTypes.arrayOf(PropTypes.string).isRequired,
  })),
  onHeaderClickHandler: PropTypes.func.isRequired,
  onFilmCardClickHandler: PropTypes.func.isRequired,
  onItemEnterHandler: PropTypes.func,
  onItemLeaveHandler: PropTypes.func,
  loading: PropTypes.func,
  activeItem: PropTypes.any,
};

const mapStateToProps = (state) => ({
  filmsList: getMyListFilms(state)
});

const mapDispatchToProps = (dispatch) => ({
  loading() {
    dispatch(DataOperation.loadMyListFilms());
  }
});

export {MyList};

export default connect(mapStateToProps, mapDispatchToProps)(MyList);
