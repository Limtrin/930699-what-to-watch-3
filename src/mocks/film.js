export const Film = {
  id: `grand-budapest`,
  name: `The Grand Budapest Hotel`,
  genre: `Dramas`,
  year: 2019,
  image: `img/bohemian-rhapsody.jpg`,
  poster: `img/the-grand-budapest-hotel-poster.jpg`,
  cover: `img/bg-the-grand-budapest-hotel.jpg`,
  preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  time: `2h 22m`,
  rating: 9.5,
  votes: 123,
  director: `Frank Spaskukotski`,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes).`,
  reviews: [
    {
      rating: 10,
      date: `March 15, 2019`,
      author: {
        id: 1,
        name: `Jane Doe`
      },
      text: `Fantastic!`
    },
    {
      rating: 10,
      date: `March 15, 2019`,
      author: {
        id: 1,
        name: `Jane Doe`
      },
      text: `Fantastic!`
    },
    {
      rating: 10,
      date: `March 15, 2019`,
      author: {
        id: 1,
        name: `Jane Doe`
      },
      text: `Fantastic!`
    }
  ],
  starring: [
    `Tim Robbins`,
    `Morgan Freeman`,
    `Bob Gunton`,
    `William Sadler`,
    `Clancy Brown`,
    `Gil Bellows`,
    `James Whitmore`
  ],
};
