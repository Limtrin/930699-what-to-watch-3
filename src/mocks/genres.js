
const FILMS_GENRES = [
  `All genres`,
  `Comedies`,
  `Crime`,
  `Documentary`,
  `Dramas`,
  `Horror`,
  `Kids & Family`,
  `Romance`,
  `Sci-Fi`,
  `Thrillers`,
];

export default FILMS_GENRES;
