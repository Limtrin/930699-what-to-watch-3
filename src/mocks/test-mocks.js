export const filmsList = [
  {
    id: `shutter-island`,
    name: `Shutter Island`,
    genre: `detective`,
    year: 2010,
    image: `img/shutter-island.jpg`,
    poster: `img/the-grand-budapest-hotel-poster.jpg`,
    cover: `img/bg-the-grand-budapest-hotel.jpg`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    time: `2h 22m`,
    rating: 9.5,
    votes: 123,
    director: `Frank Spaskukotski`,
    description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes).`,
    reviews: [
      {
        rating: 10,
        date: `March 15, 2019`,
        author: {
          id: 1,
          name: `Jane Doe`
        },
        text: `Fantastic!`
      },
      {
        rating: 10,
        date: `March 15, 2019`,
        author: {
          id: 1,
          name: `Jane Doe`
        },
        text: `Fantastic!`
      },
      {
        rating: 10,
        date: `March 15, 2019`,
        author: {
          id: 1,
          name: `Jane Doe`
        },
        text: `Fantastic!`
      }
    ],
    starring: [
      `Tim Robbins`,
      `Morgan Freeman`,
      `Bob Gunton`,
      `William Sadler`,
      `Clancy Brown`,
      `Gil Bellows`,
      `James Whitmore`
    ],
  },
  {
    id: `snatch`,
    name: `Snatch`,
    genre: `comedy`,
    year: 2009,
    image: `img/snatch.jpg`,
    poster: `img/the-grand-budapest-hotel-poster.jpg`,
    cover: `img/bg-the-grand-budapest-hotel.jpg`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    time: `2h 22m`,
    rating: 9.5,
    votes: 123,
    director: `Frank Spaskukotski`,
    description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes).`,
    reviews: [
      {
        rating: 10,
        date: `March 15, 2019`,
        author: {
          id: 1,
          name: `Jane Doe`
        },
        text: `Fantastic!`
      },
      {
        rating: 10,
        date: `March 15, 2019`,
        author: {
          id: 1,
          name: `Jane Doe`
        },
        text: `Fantastic!`
      },
      {
        rating: 10,
        date: `March 15, 2019`,
        author: {
          id: 1,
          name: `Jane Doe`
        },
        text: `Fantastic!`
      }
    ],
    starring: [
      `Tim Robbins`,
      `Morgan Freeman`,
      `Bob Gunton`,
      `William Sadler`,
      `Clancy Brown`,
      `Gil Bellows`,
      `James Whitmore`
    ],
  }
];

export const film = {
  id: `war-of-the-worlds`,
  name: `War of the Worlds`,
  genre: `action`,
  year: 2008,
  image: `img/war-of-the-worlds.jpg`,
  poster: `img/the-grand-budapest-hotel-poster.jpg`,
  cover: `img/bg-the-grand-budapest-hotel.jpg`,
  preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  time: `2h 22m`,
  rating: 9.5,
  votes: 123,
  director: `Frank Spaskukotski`,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes).`,
  reviews: [
    {
      rating: 10,
      date: `March 15, 2019`,
      author: {
        id: 1,
        name: `Jane Doe`
      },
      text: `Fantastic!`
    },
    {
      rating: 10,
      date: `March 15, 2019`,
      author: {
        id: 1,
        name: `Jane Doe`
      },
      text: `Fantastic!`
    },
    {
      rating: 10,
      date: `March 15, 2019`,
      author: {
        id: 1,
        name: `Jane Doe`
      },
      text: `Fantastic!`
    }
  ],
  starring: [
    `Tim Robbins`,
    `Morgan Freeman`,
    `Bob Gunton`,
    `William Sadler`,
    `Clancy Brown`,
    `Gil Bellows`,
    `James Whitmore`
  ],
};
