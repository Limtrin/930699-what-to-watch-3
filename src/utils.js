export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const ratingTransition = (rating) => {

  let filmRating;
  switch (true) {
    case (rating <= 3):
      filmRating = `Bad`;
      break;
    case (rating <= 5):
      filmRating = `Normal`;
      break;
    case (rating <= 8):
      filmRating = `Good`;
      break;
    case (rating < 10):
      filmRating = `Very Good`;
      break;
    default:
      filmRating = `Awesome`;
  }
  return filmRating;
};
