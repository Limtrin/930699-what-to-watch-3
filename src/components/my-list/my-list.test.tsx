import * as React from "react";
import * as renderer from "react-test-renderer";
import {MyList} from "./my-list";
import {filmsList} from "../../mocks/test-mocks";
import {BrowserRouter} from "react-router-dom";
import {noop} from "../../utils";

it(`Render MyList`, () => {
  const main = renderer
    .create(
        <BrowserRouter >
          <MyList
            onFilmCardClickHandler={noop}
            onFilmFavoriteStatusClickHandler={noop}
            onHeaderClickHandler={noop}
            loading={noop}
            onItemEnterHandler={noop}
            onItemLeaveHandler={noop}
            filmsList={filmsList}
            activeItem={null}
          />
        </BrowserRouter>, {
          createNodeMock: () => {
            return {};
          }
        }
    )
    .toJSON();

  expect(main).toMatchSnapshot();
});
