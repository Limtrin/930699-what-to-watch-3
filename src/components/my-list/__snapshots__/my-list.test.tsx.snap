// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render MyList 1`] = `
<div
  className="user-page"
>
  <header
    className="page-header user-page__head"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <h1
      className="page-title user-page__title"
    >
      My list
    </h1>
    <div
      className="user-block"
    >
      <div
        className="user-block__avatar"
      >
        <img
          alt="User avatar"
          height="63"
          src="img/avatar.jpg"
          width="63"
        />
      </div>
    </div>
  </header>
  <section
    className="catalog"
  >
    <h2
      className="catalog__title visually-hidden"
    >
      Catalog
    </h2>
    <div
      className="catalog__movies-list"
    >
      <article
        className="small-movie-card catalog__movies-card"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Pulp Fiction"
            height="175"
            src="img/shutter-island.jpg"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/shutter-island"
            onClick={[Function]}
          >
            Shutter Island
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Pulp Fiction"
            height="175"
            src="img/snatch.jpg"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/snatch"
            onClick={[Function]}
          >
            Snatch
          </a>
        </h3>
      </article>
    </div>
  </section>
  <footer
    className="page-footer"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link logo__link--light"
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="copyright"
    >
      <p>
        Â© 2019 What to watch Ltd.
      </p>
    </div>
  </footer>
</div>
`;
